next_sparseconvnet/networks/architectures.py

```diff
@@ -147,3 +147,118 @@ def forward(self, x):
+        '''
+        Parameters
+        ----------
+        spatial_size : tuple
+            The spatial size of the input layer. Size of the tuple is also the dimension.
+        init_conv_nplaness : int
+            Number of planes we want after the initial SubmanifoldConvolution, that is, to begin downsampling.
+        init_conv_kernel : int
+            Kernel for the first convolutional layer.
+        kernel_sizes : list
+            Size of the kernels applied in each layer or block. Its length also indicates the level depth of the net.
+            Last element corresponds just to the kernel of the basic block in the bottom of the net.
+        stride : int
+            Applied stride in every layer or block.
+        basic_num : int
+            Number of times a basic residual block is passed in each level.
+        nclasses : int, optional
+            Number of output classes for predictions. The default is 2.
+        dim : int, optional
+            Number of dimensions of the input. The default is 3.
+        start_planes : int, optional
+            Number of planes that enter the net. The default is 1.
+        '''
+        torch.nn.Module.__init__(self)
+
+        self.basic_num   = basic_num
+        self.level_depth = len(kernel_sizes)
+
+        self.inp     = scn.InputLayer(dim, spatial_size)
+        self.convBN  = ConvBNBlock(start_planes, init_conv_nplanes, init_conv_kernel)
+        inplanes     = init_conv_nplanes
+        self.max     = scn.MaxPooling(dim,2,2)
+        self.sparse  = scn.SparseToDense(dim,inplanes)
+        self.linear  = torch.nn.Linear(inplanes, nclasses)
+
+        self.downsample = torch.nn.ModuleList([])
+        self.basic      = torch.nn.ModuleList([torch.nn.ModuleList([]) for i in range(self.level_depth - 1)])
+        self.bottom     = torch.nn.ModuleList([])
+
+        for i in range(self.level_depth - 1):
+            for j in range(basic_num):
+                self.basic[i].append(ResidualBlock_basic(inplanes, kernel_sizes[i]))
+
+                self.downsample.append(ResidualBlock_downsample(inplanes, kernel_sizes[i], stride))
+
+                inplanes = inplanes * 2
+
+    def forward(self, x):
+    '''
+    Passes x through the net.
+
+    Parameters
+    ----------
+    x : tuple
+        It takes a tuple with (coord, features), where coord is a torch tensor with size [features_number, batch_size],
+        and features is another torch tensor, with size [features_number, start_planes]
+
+    Returns
+    -------
+    x : torch.Tensor
+        A tensor with size [features_number, nclasses].
+
+    '''
+    x = self.inp(x)
+    x = self.convBN(x)
+
+    for i in range(self.level_depth - 1):
+        for j in range(self.basic_num):
+            x = self.basic[i][j](x)
+        x = self.downsample[i](x)
+
+    for i in range(self.basic_num):
+        x = self.bottom[i](x)
+
+    x = self.max(x)
+    x = self.sparse(x)
```

Comment:
This layer will give output of 
`batch_size, outplanes, spatial_x_dim, spatial_y_dim, spatial_z_dim`,
in case you calculate properly max pool size so you preformed global pool the spatial dimensions will be ==1 in which case you could call `x=x.squeeze()` to simply remove them (after the `self.sparse` layer).
In case you havent performed global pool you need to add a Flatten layer